scripts/git_fix.py

```diff
@@ -0,0 +1,213 @@
+
+  return {f for f in files if os.path.splitext(f)[1] == '.py'}
+
+
+def tool_is_installed(file_type: str, tool: str) -> bool:
+  """Checks if the given tool is installed.
+
+  Args:
+    file_type: The type of the file (e.g. C++, python).
+    tool: The name of the tool.
+
+  Returns:
+    True if the tool is already installed.  
+  """
+
+  # Is the tool already installed?
+  if os.system(f'which {tool} > /dev/null') == 0:
+    return True
+
+  print(
+      f'WARNING: {tool} is not installed. It\'s needed for formatting '
+      f'{file_type} files. Please install it and retry.',
+      file=sys.stderr)
+  return False
+
+
+def format_files(file_type: str, tool: str, format_command: str,
+                 files: set[str]) -> bool:
+  """Fixes the formatting of the given files.
+  
+  Args:
+    file_type: The type of the file (e.g. C++, python).    
+    tool: The name of the tool.
+    format_command: The command to format the files.
+    files: The files to format.
+  Returns:
+    True if the formatting was successful.
+  """
+
+  if not files:
+    return True
+
+  if not tool_is_installed(file_type, tool):
+    return False
+
+  command = f'{format_command} {" ".join(files)}'
+  if os.system(command) == 0:
+    print(
+        f'Successfully formatted {file_type} files:\n' +
+        '\n'.join(sorted('  ' + f for f in files)),
+        file=sys.stderr)
+    return True
+
+  print(
+      f'WARNING: Failed to format {file_type} files via command: {command}',
+      file=sys.stderr)
+  return False
+
+
+def format_cplusplus_files(files: set[str]) -> bool:
+  """Fixes the formatting of the given C++ files.
+  
+  Returns:
+    True if the formatting was successful.
+  """
+
+  return format_files('C++', _CLANG_FORMAT, f'{_CLANG_FORMAT} -i -style=file',
+                      files)
+
+
+def format_python_files(files: set[str]) -> bool:
+  """Fixes the formatting of the given Python files.
+
+  Returns:
+    True if the formatting was successful.
+  """
+
+  return format_files('python', _YAPF, f'{_YAPF} -i', files)
+
+
+def set_git_push_hook() -> bool:
+  """Sets up `git push` to automatially run this script before pushing.
+  
+  Returns:
+    True if the set-up was successful.
+  """
+
+  if os.path.isfile(_GIT_PRE_PUSH_HOOK_PATH):
+    print(
+        f'git pre-push hook already exists at {_GIT_PRE_PUSH_HOOK_PATH}.',
+        file=sys.stderr)
+    # Ask the user if they want to overwrite it.
+    overwrite = input(
+        f'Are you sure you want to overwrite {_GIT_PRE_PUSH_HOOK_PATH}? [y/N] ')
+    if overwrite.lower() != 'y':
+      print('Skipping git pre-push hook setup.', file=sys.stderr)
+      return False
+
+  # Create a git pre-push hook to run this script.
+  with open(_GIT_PRE_PUSH_HOOK_PATH, 'w') as f:
+    f.write(f'''#!/bin/bash
+# This hook is automatically set by `{_SCRIPT_PATH } --set_git_push_hook`.)
+# We ignore any errors, as file formatting is best effort.
```

Comment:
Should we abort the push operation from inside the hook if the files are formatted? Seems useful since otherwise the user might forget that they need to re-commit the formatted files and push again.

Reply:
I decided not to block on formatting failures as the formatting could fail due to the user not having the proper tools installed. In this case, we should let the user push and rely on feedback from CI to format the files manually.

As a future improvement, we can make the script check if the formatter made any changes, and abort if yes so that the user will know they need to commit the changes.

Thanks!